pkg/service/well_known_attributes.go

```diff
@@ -0,0 +1,201 @@
+	// Downstream connection local address
+	Address string `json:"address,omitempty"`
+	// Downstream connection local port e.g. 9090
+	Port int32 `json:"port,omitempty"`
+	// The canonical service name of the peer e.g. “foo.default.svc.cluster.local”
+	Service string `json:"service,omitempty"`
+	// The labels associated with the peer. These could be pod labels for Kubernetes or tags for VMs. The source of the
+	// labels could be an X.509 certificate or other configuration.
+	Labels map[string]string `json:"labels,omitempty"`
+	// The authenticated identity of this peer. If an X.509 certificate is used to assert the identity in the proxy, this
+	// field is sourced from "URI Subject Alternative Names", "DNS Subject Alternate Names" or "Subject" in that order.
+	// The format is issuer specific – e.g. SPIFFE format is spiffe://trust-domain/path, Google account format is https://accounts.google.com/{userid}.
+	Principal string `json:"principal,omitempty"`
+	// The X.509 certificate used to authenticate the identity of this peer. When present, the certificate contents are encoded in URL and PEM format.
+	Certificate string `json:"certificate,omitempty"`
+}
+
+type AuthAttributes struct {
+	// Single resolved identity object, post-identity verification
+	Identity any `json:"identity,omitempty"`
+	// External metadata fetched
+	Metadata map[string]any `json:"metadata,omitempty"`
+	// Authorization results resolved by each authorization rule, access granted only
+	Authorization map[string]any `json:"authorization,omitempty"`
+	// Response objects exported by the auth service post-access granted
+	Response map[string]any `json:"response,omitempty"`
+	// Response objects returned by the callback requests issued by the auth service
+	Callbacks map[string]any `json:"callbacks,omitempty"`
+}
+
+// NewWellKnownAttributes creates a new WellKnownAttributes object from an envoyauth.AttributeContext
+func NewWellKnownAttributes(attributes *envoyauth.AttributeContext, authData map[string]any) *WellKnownAttributes {
+	return &WellKnownAttributes{
+		Metadata:    attributes.MetadataContext,
+		Request:     newRequestAttributes(attributes),
+		Source:      newSourceAttributes(attributes),
+		Destination: newDestinationAttributes(attributes),
+		Auth:        newAuthAttributes(authData),
+	}
+}
+
+func newRequestAttributes(attributes *envoyauth.AttributeContext) *RequestAttributes {
+	request := attributes.GetRequest()
+	httpRequest := request.GetHttp()
+	urlParsed, _ := url.Parse(httpRequest.Path)
+	headers := httpRequest.GetHeaders()
+	return &RequestAttributes{
+		Id:                httpRequest.Id,
+		Time:              request.Time,
+		Protocol:          httpRequest.Protocol,
+		Scheme:            httpRequest.GetScheme(),
+		Host:              httpRequest.GetHost(),
+		Method:            httpRequest.GetMethod(),
+		Path:              httpRequest.GetPath(),
+		URLPath:           urlParsed.Path,
+		Query:             httpRequest.GetQuery(),
```

Comment:
I'm afraid we might need to implement this one ourselves: https://github.com/envoyproxy/go-control-plane/blob/1d7ad182c6f4820d713680e3987538e28917e62b/envoy/service/auth/v3/attribute_context.pb.go#L324-L326